# Result with all tags

In [1]:
from prettytable import PrettyTable
print("models with all tags")
tabel = PrettyTable()
tabel.field_names=['Model','Vectorizer','analyzer','ngrams','Precision','recall','f1_score','hyperparameter']
tabel.add_row(['MultinomialNB','CountVectorizer','word','1,3','0.417','0.315','0.359','1'])
tabel.add_row(['SGDClassifier(log)','CountVectorizer','word','1,3','0.414','0.251','0.313','0.01'])
tabel.add_row(['SGDClassifier(hinge)','CountVectorizer','word','1,3','0.359','0.239','0.287','0.01'])
tabel.add_row(['MultinomialNB','TfidfVectorizer','word','1,3','0.585','0.149','0.238','0.01'])
tabel.add_row(['SGDClassifier(log)','TfidfVectorizer','word','1,3','0.437','0.222','0.295','1e-06'])
tabel.add_row(['SGDClassifier(hinge)','TfidfVectorizer','word','1,3','0.413','0.235','0.300','1e-05'])
tabel.add_row(['MultinomialNB','CountVectorizer','char','3,4','0.354','0.343','0.349','2'])
tabel.add_row(['SGDClassifier(log)','CountVectorizer','char','3,4','0.354','0.236','0.283','0.1'])
tabel.add_row(['SGDClassifier(hinge)','CountVectorizer','char','3,4','0.262','0.236','0.248','0.01'])
tabel.add_row(['MultinomialNB','TfidfVectorizer','char','3,4','0.539','0.150','0.235','0.01'])
tabel.add_row(['SGDClassifier(log)','TfidfVectorizer','char','3,4','0.404','0.235','0.295','1e-06'])
tabel.add_row(['SGDClassifier(hinge)','TfidfVectorizer','char','3,4','0.415','0.228','0.294','1e-05'])
print(tabel)

models with all tags
+----------------------+-----------------+----------+--------+-----------+--------+----------+----------------+
|        Model         |    Vectorizer   | analyzer | ngrams | Precision | recall | f1_score | hyperparameter |
+----------------------+-----------------+----------+--------+-----------+--------+----------+----------------+
|    MultinomialNB     | CountVectorizer |   word   |  1,3   |   0.417   | 0.315  |  0.359   |       1        |
|  SGDClassifier(log)  | CountVectorizer |   word   |  1,3   |   0.414   | 0.251  |  0.313   |      0.01      |
| SGDClassifier(hinge) | CountVectorizer |   word   |  1,3   |   0.359   | 0.239  |  0.287   |      0.01      |
|    MultinomialNB     | TfidfVectorizer |   word   |  1,3   |   0.585   | 0.149  |  0.238   |      0.01      |
|  SGDClassifier(log)  | TfidfVectorizer |   word   |  1,3   |   0.437   | 0.222  |  0.295   |     1e-06      |
| SGDClassifier(hinge) | TfidfVectorizer |   word   |  1,3   |   0.413   | 0.235  |

# Results with top5 tags 

In [2]:
from prettytable import PrettyTable
table = PrettyTable()
print("CountVectorization with word analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','1,1','0.497','1'])
table.add_row(['SGDClassifier(log)','1,1','0.444','0.01'])
table.add_row(['SGDClassifier(hinge)','1,1','0.441','0.1'])
table.add_row(['MultinomialNB','2,2','0.501','1'])
table.add_row(['SGDClassifier(log)','2,2','0.420','0.0001'])
table.add_row(['SGDClassifier(hinge)','2,2','0.422','0.001'])
table.add_row(['MultinomialNB','3,3','0.399','1'])
table.add_row(['SGDClassifier(log)','3,3','0.290','1e-05'])
table.add_row(['SGDClassifier(hinge)','3,3','0.313','1e-05'])
table.add_row(['MultinomialNB','1,2','0.52','1'])
table.add_row(['SGDClassifier(log)','1,2','0.445','0.01'])
table.add_row(['SGDClassifier(hinge)','1,2','0.452','0.1'])
table.add_row(['MultinomialNB','1,3','0.52','1'])
table.add_row(['SGDClassifier(log)','1,3','0.465','0.1'])
table.add_row(['SGDClassifier(hinge)','1,3','0.463','0.1'])
print(table)
print("")
table = PrettyTable()
print("CountVectorization with char analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','3,3','0.500','2.5'])
table.add_row(['SGDClassifier(log)','3,3','0.468','1'])
table.add_row(['SGDClassifier(hinge)','3,3','0.435','0.0001'])
table.add_row(['MultinomialNB','4,4','0.506','2'])
table.add_row(['SGDClassifier(log)','4,4','0.440','0.1'])
table.add_row(['SGDClassifier(hinge)','4,4','0.469','0.1'])
table.add_row(['MultinomialNB','3,4','0.502','1'])
table.add_row(['SGDClassifier(log)','3,4','0.460','1'])
table.add_row(['SGDClassifier(hinge)','3,4','0.475','0.1'])
table.add_row(['MultinomialNB','2,4','0.503','1'])
table.add_row(['SGDClassifier(log)','2,4','0.388','1e-05'])
table.add_row(['SGDClassifier(hinge)','2,4','0.418','0.1'])
table.add_row(['MultinomialNB','1,4','0.502','1'])
table.add_row(['SGDClassifier(log)','1,4','0.465','0.1'])
table.add_row(['SGDClassifier(hinge)','1,4','0.370','2'])
print(table)

CountVectorization with word analyzer
+----------------------+--------+----------+----------------+
|        Model         | ngrams | f1_score | hyperparameter |
+----------------------+--------+----------+----------------+
|    MultinomialNB     |  1,1   |  0.497   |       1        |
|  SGDClassifier(log)  |  1,1   |  0.444   |      0.01      |
| SGDClassifier(hinge) |  1,1   |  0.441   |      0.1       |
|    MultinomialNB     |  2,2   |  0.501   |       1        |
|  SGDClassifier(log)  |  2,2   |  0.420   |     0.0001     |
| SGDClassifier(hinge) |  2,2   |  0.422   |     0.001      |
|    MultinomialNB     |  3,3   |  0.399   |       1        |
|  SGDClassifier(log)  |  3,3   |  0.290   |     1e-05      |
| SGDClassifier(hinge) |  3,3   |  0.313   |     1e-05      |
|    MultinomialNB     |  1,2   |   0.52   |       1        |
|  SGDClassifier(log)  |  1,2   |  0.445   |      0.01      |
| SGDClassifier(hinge) |  1,2   |  0.452   |      0.1       |
|    MultinomialNB     |  1,3   

In [3]:
from prettytable import PrettyTable
table = PrettyTable()
print("TfidfVectorization with word analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','1,1','0.364','0.01'])
table.add_row(['SGDClassifier(log)','1,1','0.442','1e-06'])
table.add_row(['SGDClassifier(hinge)','1,1','0.452','1e-06'])
table.add_row(['MultinomialNB','2,2','0.420','0.1'])
table.add_row(['SGDClassifier(log)','2,2','0.386','1e-06'])
table.add_row(['SGDClassifier(hinge)','2,2','0.424','1e-06'])
table.add_row(['MultinomialNB','3,3','0.272','0.01'])
table.add_row(['SGDClassifier(log)','3,3','0.344','1e-06'])
table.add_row(['SGDClassifier(hinge)','3,3','0.333','1e-06'])
table.add_row(['MultinomialNB','1,2','0.415','0.01'])
table.add_row(['SGDClassifier(log)','1,2','0.446','1e-06'])
table.add_row(['SGDClassifier(hinge)','1,2','0.454','1e-06'])
table.add_row(['MultinomialNB','1,3','0.417','0.01'])
table.add_row(['SGDClassifier(log)','1,3','0.445','1e-06'])
table.add_row(['SGDClassifier(hinge)','1,3','0.451','1e-05'])
table.add_row(['MultinomialNB','1,4','0.417','0.01'])
table.add_row(['SGDClassifier(log)','1,4','0.456','1e-06'])
table.add_row(['SGDClassifier(hinge)','1,4','0.451','1e-06'])
print(table)
print('')
table = PrettyTable()
print("TfidfVectorization with char analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','3,3','0.226','0.001'])
table.add_row(['SGDClassifier(log)','3,3','0.466','1e-05'])
table.add_row(['SGDClassifier(hinge)','3,3','0.429','1e-05'])
table.add_row(['MultinomialNB','4,4','0.384','0.01'])
table.add_row(['SGDClassifier(log)','4,4','0.431','1e-06'])
table.add_row(['SGDClassifier(hinge)','4,4','0.432','1e-05'])
table.add_row(['MultinomialNB','3,4','0.393','0.01'])
table.add_row(['SGDClassifier(log)','3,4','0.442','1e-05'])
table.add_row(['SGDClassifier(hinge)','3,4','0.447','1e-08'])
table.add_row(['MultinomialNB','2,4','0.373','0.01'])
table.add_row(['SGDClassifier(log)','2,4','0.455','1e-05'])
table.add_row(['SGDClassifier(hinge)','2,4','0.440','1e-05'])
table.add_row(['MultinomialNB','1,4','0.283','0.001'])
table.add_row(['SGDClassifier(log)','1,4','0.448','1e-05'])
table.add_row(['SGDClassifier(hinge)','1,4','0.434','1e-06'])
print(table)

TfidfVectorization with word analyzer
+----------------------+--------+----------+----------------+
|        Model         | ngrams | f1_score | hyperparameter |
+----------------------+--------+----------+----------------+
|    MultinomialNB     |  1,1   |  0.364   |      0.01      |
|  SGDClassifier(log)  |  1,1   |  0.442   |     1e-06      |
| SGDClassifier(hinge) |  1,1   |  0.452   |     1e-06      |
|    MultinomialNB     |  2,2   |  0.420   |      0.1       |
|  SGDClassifier(log)  |  2,2   |  0.386   |     1e-06      |
| SGDClassifier(hinge) |  2,2   |  0.424   |     1e-06      |
|    MultinomialNB     |  3,3   |  0.272   |      0.01      |
|  SGDClassifier(log)  |  3,3   |  0.344   |     1e-06      |
| SGDClassifier(hinge) |  3,3   |  0.333   |     1e-06      |
|    MultinomialNB     |  1,2   |  0.415   |      0.01      |
|  SGDClassifier(log)  |  1,2   |  0.446   |     1e-06      |
| SGDClassifier(hinge) |  1,2   |  0.454   |     1e-06      |
|    MultinomialNB     |  1,3   

# Results with top3 tags

In [4]:
from prettytable import PrettyTable
table = PrettyTable()
print("CountVectorization with word analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','1,1','0.535','1'])
table.add_row(['SGDClassifier(log)','1,1','0.504','0.1'])
table.add_row(['SGDClassifier(hinge)','1,1','0.507','0.1'])
table.add_row(['MultinomialNB','2,2','0.547','2'])
table.add_row(['SGDClassifier(log)','2,2','0.453','1e-05'])
table.add_row(['SGDClassifier(hinge)','2,2','0.479','1e-05'])
table.add_row(['MultinomialNB','3,3','0.384','1'])
table.add_row(['SGDClassifier(log)','3,3','0.375','1e-05'])
table.add_row(['SGDClassifier(hinge)','3,3','0.384','1e-05'])
table.add_row(['MultinomialNB','1,2','0.560','1'])
table.add_row(['SGDClassifier(log)','1,2','0.511','0.1'])
table.add_row(['SGDClassifier(hinge)','1,2','0.505','0.1'])
table.add_row(['MultinomialNB','1,3','0.560','1'])
table.add_row(['SGDClassifier(log)','1,3','0.496','0.01'])
table.add_row(['SGDClassifier(hinge)','1,3','0.486','0.1'])
print(table)
print("")
table = PrettyTable()
print("CountVectorization with char analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','3,3','0.536','2.5'])
table.add_row(['SGDClassifier(log)','3,3','0.438','2.5'])
table.add_row(['SGDClassifier(hinge)','3,3','0.493','0.0001'])
table.add_row(['MultinomialNB','4,4','0.546','2'])
table.add_row(['SGDClassifier(log)','4,4','0.539','0.1'])
table.add_row(['SGDClassifier(hinge)','4,4','0.477','0.1'])
table.add_row(['MultinomialNB','3,4','0.542','2'])
table.add_row(['SGDClassifier(log)','3,4','0.521','1'])
table.add_row(['SGDClassifier(hinge)','3,4','0.510','1'])
table.add_row(['MultinomialNB','2,4','0.541','1'])
table.add_row(['SGDClassifier(log)','2,4','0.463','1'])
table.add_row(['SGDClassifier(hinge)','2,4','0.434','2.5'])
table.add_row(['MultinomialNB','1,4','0.541','1'])
table.add_row(['SGDClassifier(log)','1,4','0.511','1'])
table.add_row(['SGDClassifier(hinge)','1,4','0.538','0.001'])
print(table)

CountVectorization with word analyzer
+----------------------+--------+----------+----------------+
|        Model         | ngrams | f1_score | hyperparameter |
+----------------------+--------+----------+----------------+
|    MultinomialNB     |  1,1   |  0.535   |       1        |
|  SGDClassifier(log)  |  1,1   |  0.504   |      0.1       |
| SGDClassifier(hinge) |  1,1   |  0.507   |      0.1       |
|    MultinomialNB     |  2,2   |  0.547   |       2        |
|  SGDClassifier(log)  |  2,2   |  0.453   |     1e-05      |
| SGDClassifier(hinge) |  2,2   |  0.479   |     1e-05      |
|    MultinomialNB     |  3,3   |  0.384   |       1        |
|  SGDClassifier(log)  |  3,3   |  0.375   |     1e-05      |
| SGDClassifier(hinge) |  3,3   |  0.384   |     1e-05      |
|    MultinomialNB     |  1,2   |  0.560   |       1        |
|  SGDClassifier(log)  |  1,2   |  0.511   |      0.1       |
| SGDClassifier(hinge) |  1,2   |  0.505   |      0.1       |
|    MultinomialNB     |  1,3   

In [5]:
from prettytable import PrettyTable
table = PrettyTable()
print("TfidfVectorization with word analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','1,1','0.416','0.01'])
table.add_row(['SGDClassifier(log)','1,1','0.517','1e-05'])
table.add_row(['SGDClassifier(hinge)','1,1','0.498','1e-05'])
table.add_row(['MultinomialNB','2,2','0.490','0.1'])
table.add_row(['SGDClassifier(log)','2,2','0.463','1e-06'])
table.add_row(['SGDClassifier(hinge)','2,2','0.475','1e-06'])
table.add_row(['MultinomialNB','3,3','0.321','0.01'])
table.add_row(['SGDClassifier(log)','3,3','0.377','1e-06'])
table.add_row(['SGDClassifier(hinge)','3,3','0.361','1e-06'])
table.add_row(['MultinomialNB','1,2','0.476','0.01'])
table.add_row(['SGDClassifier(log)','1,2','0.497','1e-05'])
table.add_row(['SGDClassifier(hinge)','1,2','0.486','1e-06'])
table.add_row(['MultinomialNB','1,3','0.480','0.01'])
table.add_row(['SGDClassifier(log)','1,3','0.500','1e-05'])
table.add_row(['SGDClassifier(hinge)','1,3','0.503','1e-05'])
table.add_row(['MultinomialNB','1,4','0.480','0.01'])
table.add_row(['SGDClassifier(log)','1,4','0.501','1e-06'])
table.add_row(['SGDClassifier(hinge)','1,4','0.503','1e-05'])
print(table)
print('')
table = PrettyTable()
print("TfidfVectorization with char analyzer")
table.field_names=['Model','ngrams','f1_score','hyperparameter']
table.add_row(['MultinomialNB','3,3','0.283','0.0001'])
table.add_row(['SGDClassifier(log)','3,3','0.540','1e-05'])
table.add_row(['SGDClassifier(hinge)','3,3','0.504','1e-05'])
table.add_row(['MultinomialNB','4,4','0.443','0.01'])
table.add_row(['SGDClassifier(log)','4,4','0.502','1e-05'])
table.add_row(['SGDClassifier(hinge)','4,4','0.471','1e-07'])
table.add_row(['MultinomialNB','3,4','0.451','0.01'])
table.add_row(['SGDClassifier(log)','3,4','0.509','1e-05'])
table.add_row(['SGDClassifier(hinge)','3,4','0.500','1e-09'])
table.add_row(['MultinomialNB','2,4','0.435','0.01'])
table.add_row(['SGDClassifier(log)','2,4','0.435','0.01'])
table.add_row(['SGDClassifier(hinge)','2,4','0.500','1e-07'])
table.add_row(['MultinomialNB','1,4','0.283','0.001'])
table.add_row(['SGDClassifier(log)','1,4','0.346','1e-05'])
table.add_row(['SGDClassifier(hinge)','1,4','0.429','1e-07'])
print(table)

TfidfVectorization with word analyzer
+----------------------+--------+----------+----------------+
|        Model         | ngrams | f1_score | hyperparameter |
+----------------------+--------+----------+----------------+
|    MultinomialNB     |  1,1   |  0.416   |      0.01      |
|  SGDClassifier(log)  |  1,1   |  0.517   |     1e-05      |
| SGDClassifier(hinge) |  1,1   |  0.498   |     1e-05      |
|    MultinomialNB     |  2,2   |  0.490   |      0.1       |
|  SGDClassifier(log)  |  2,2   |  0.463   |     1e-06      |
| SGDClassifier(hinge) |  2,2   |  0.475   |     1e-06      |
|    MultinomialNB     |  3,3   |  0.321   |      0.01      |
|  SGDClassifier(log)  |  3,3   |  0.377   |     1e-06      |
| SGDClassifier(hinge) |  3,3   |  0.361   |     1e-06      |
|    MultinomialNB     |  1,2   |  0.476   |      0.01      |
|  SGDClassifier(log)  |  1,2   |  0.497   |     1e-05      |
| SGDClassifier(hinge) |  1,2   |  0.486   |     1e-06      |
|    MultinomialNB     |  1,3   

# Glove

In [6]:
table = PrettyTable()
print("Using all the tags")
table.field_names=['model','score','hyperparameter']
table.add_row(['SGDClassifier(log)','0.230','1e-06'])
table.add_row(['SGDClassifier(hinge)','0.087','1e-05'])
print(table)

table = PrettyTable()
print("Using top 5 tags")
table.field_names=['model','score','hyperparameter']
table.add_row(['SGDClassifier(log)','0.295','1e-07'])
table.add_row(['SGDClassifier(hinge)','0.416','1e-07'])
print(table)

table = PrettyTable()
print("Using top 3 tags")
table.field_names=['model','score','hyperparameter']
table.add_row(['SGDClassifier(log)','0.473','1e-07'])
table.add_row(['SGDClassifier(hinge)','0.474','1e-11'])
print(table)

Using all the tags
+----------------------+-------+----------------+
|        model         | score | hyperparameter |
+----------------------+-------+----------------+
|  SGDClassifier(log)  | 0.230 |     1e-06      |
| SGDClassifier(hinge) | 0.087 |     1e-05      |
+----------------------+-------+----------------+
Using top 5 tags
+----------------------+-------+----------------+
|        model         | score | hyperparameter |
+----------------------+-------+----------------+
|  SGDClassifier(log)  | 0.295 |     1e-07      |
| SGDClassifier(hinge) | 0.416 |     1e-07      |
+----------------------+-------+----------------+
Using top 3 tags
+----------------------+-------+----------------+
|        model         | score | hyperparameter |
+----------------------+-------+----------------+
|  SGDClassifier(log)  | 0.473 |     1e-07      |
| SGDClassifier(hinge) | 0.474 |     1e-11      |
+----------------------+-------+----------------+


# CONCLUSION

1. MultinomialNB with ngram_range(1,3) is giving the best f1_score 0.560. 
2. This score is far better than the results shown in the research paper.

# SUMMARY

1. Load the data and check for the duplicate titles and remove them.
2. Perform EDA on tags where i found their are 71 unique tags.
3. Most occuring tags are murder, violence, clut, flachback.
4. As per mentioned in the research paper i had chosen micro-f score as a metric.
5. Next step is clean the Data, lemtize the word and stem it.
6. I have performed the modeling in four different forms.
    
    i.  modeling with all the tags.
    
    ii. modeling with top5 tags.
    
    iii.modeling with top3 tags.
    
    iv. modeling wiht glove vectors.
    
7. The models which i have used are multinomialNB,ligistic regression, linearSVC.
8. MultinomialNB is turned out to be the best model with 0.56 score.